src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs

```diff
@@ -21,6 +21,278 @@ public class AttributeTests : WellKnownAttributesTestBase
+// but C won't exist
+";
+
+            var reference_cs =
+@"using System;
+public class RefersToLibAttribute : Attribute
+{
+    public RefersToLibAttribute() { }
+    public RefersToLibAttribute(C c) { }
+}
+";
+
+            var origLibComp = CreateStandardCompilation(origLib_cs, assemblyName: "lib");
+            origLibComp.VerifyDiagnostics();
+
+            var compWithReferenceToLib = CreateStandardCompilation(reference_cs, references: new[] { origLibComp.EmitToImageReference() });
+            compWithReferenceToLib.VerifyDiagnostics();
+            var imageWithReferenceToLib = compWithReferenceToLib.EmitToImageReference();
+
+            var newLibComp = CreateStandardCompilation(newLib_cs, references: new[] { imageWithReferenceToLib }, assemblyName: "lib");
+            newLibComp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(21194, "https://github.com/dotnet/roslyn/issues/21194")]
+        public void CircularTypeReferenceNotFound_WithRelevantType()
+        {
+            var origLib_cs = @"public class C : System.Attribute { }";
+
+            var newLib_cs = @"
+[assembly: RefersToLib] // to bind this, we'll need to find type C in 'lib'
+// but C won't exist
+";
+
+            var reference_cs =
+@"
+public class RefersToLibAttribute : C
+{
+    public RefersToLibAttribute() { }
+}
+";
+
+            var origLibComp = CreateStandardCompilation(origLib_cs, assemblyName: "lib");
+            origLibComp.VerifyDiagnostics();
+
+            var compWithReferenceToLib = CreateStandardCompilation(reference_cs, references: new[] { origLibComp.EmitToImageReference() });
+            compWithReferenceToLib.VerifyDiagnostics();
+            var imageWithReferenceToLib = compWithReferenceToLib.EmitToImageReference();
+
+            var newLibComp = CreateStandardCompilation(newLib_cs, references: new[] { imageWithReferenceToLib }, assemblyName: "lib");
+            newLibComp.VerifyDiagnostics(
+                // (2,12): error CS7068: Reference to type 'C' claims it is defined in this assembly, but it is not defined in source or any added modules
+                // [assembly: RefersToLib] // to bind this, we'll want to lookup type C in 'lib'
+                Diagnostic(ErrorCode.ERR_MissingTypeInSource, "RefersToLib").WithArguments("C").WithLocation(2, 12)
+                );
+        }
+
+        [Fact]
+        [WorkItem(21194, "https://github.com/dotnet/roslyn/issues/21194")]
+        public void CircularTypeReferenceNotFound_WithStaticUsing()
+        {
+            var origLib_cs = @"public class C : System.Attribute { }";
+
+            var newLib_cs = @"
+using static RefersToLibAttribute;
+    // Binding this will cause a lookup for 'C' in 'lib'.
+    // Such lookup requires binding 'TypeForwardedTo' attributes (and aliases), but we should not bind other usings, lest we want an infinite recursion.
```

Comment:
Typo?

Reply:
Sorry, "lest" is arcane. Similar meaning to "unless". I'll change

---
In reply to: [155961998](https://github.com/dotnet/roslyn/pull/23647#discussion_r155961998) [](ancestors = 155961998)